File: 2021/day14.cpp
// https://adventofcode.com/2021/day/14

#include "../common/common.h"
#include "../common/rust.h"

#include <algorithm>
#include <array>
#include <print>
#include <ranges>
#include <string>

using Polymer = String;
using Rules = std::array<std::array<char, 26>, 26>;
using Input = std::pair<Polymer, Rules>;

auto parse(String const& filename) -> Input {
  auto polymer = String{};
  auto rules = Rules{};
  auto parsing_rules = false;
  for (str line : aoc::views::read_lines(filename, aoc::keep_empty{})) {
    if (line.empty()) {
      parsing_rules = true;
    } else if (parsing_rules) {
      let[pair, insert] = aoc::split_once(line, " -> ");
      rules[pair[0] - 'A'][pair[1] - 'A'] = insert[0];
    } else {
      polymer = String{line};
    }
  }
  return {std::move(polymer), rules};
}

using Pairs = std::array<std::array<usize, 26>, 26>;

fn apply_rules(Pairs const& pairs, Rules const& rules) -> Pairs {
  auto new_pairs = Pairs{};
  for (usize a = 0; a < 26; ++a) {
    for (usize b = 0; b < 26; ++b) {
      let count = pairs[a][b];
      if (count == 0) {
        continue;
      }
      let insert = rules[a][b];
      if (insert == '\0') {
        continue;
      }
      let c = static_cast<usize>(insert - 'A');
      new_pairs[a][c] += count;
      new_pairs[c][b] += count;
    }
  }
  return new_pairs;
}

template <usize STEPS>
fn solve_case(Input const& input) -> usize {
  let & [ polymer, rules ] = input;
  auto pairs = Pairs{};
  for (usize i = 0; i + 1 < polymer.size(); ++i) {
    ++pairs[polymer[i] - 'A'][polymer[i + 1] - 'A'];
  }
  for (usize step = 0; step < STEPS; ++step) {
    pairs = apply_rules(pairs, rules);
  }
  auto counts = std::array<usize, 26>{};
  for (usize a = 0; a < 26; ++a) {
    for (usize b = 0; b < 26; ++b) {
      counts[a] += pairs[a][b];
    }
  }
  counts[polymer.back() - 'A'] += 1;
  let max = *std::ranges::max_element(counts);
  auto nonzero = counts | std::views::filter([](usize c) { return c > 0; });
  let min = *std::ranges::min_element(nonzero);
  return max - min;
}

int main() {
  std::println("Part 1");
  let example = parse("day14.example");
  AOC_EXPECT_RESULT(1588, solve_case<10>(example));
  let input = parse("day14.input");
  AOC_EXPECT_RESULT(2703, solve_case<10>(input));

  std::println("Part 2");
  AOC_EXPECT_RESULT(2188189693529, solve_case<40>(example));
  AOC_EXPECT_RESULT(2984946368465, solve_case<40>(input));

  AOC_RETURN_CHECK_RESULT();
}

File: rust/2021/src/day14.rs
use std::collections::HashMap;

type Polymer = String;
type Rules = HashMap<(char, char), char>;
type Input = (Polymer, Rules);

fn parse(filename: &str) -> Input {
    let mut polymer = String::new();
    let mut rules = HashMap::new();
    let mut parsing_rules = false;
    for line in aoc::file::read_lines(filename) {
        if line.is_empty() {
            parsing_rules = true;
        } else if parsing_rules {
            let (pair, insert) = line.split_once(" -> ").unwrap();
            let mut pair_chars = pair.chars();
            let a = pair_chars.next().unwrap();
            let b = pair_chars.next().unwrap();
            let c = insert.chars().next().unwrap();
            rules.insert((a, b), c);
        } else {
            polymer = line;
        }
    }
    (polymer, rules)
}

type Pairs = HashMap<(char, char), usize>;

fn apply_rules(pairs: &Pairs, rules: &Rules) -> Pairs {
    let mut new_pairs = Pairs::new();
    for (rule_pair, insert) in rules {
        match pairs.get(rule_pair) {
            None => {}
            Some(count) => {
                *new_pairs.entry((rule_pair.0, *insert)).or_insert(0) += count;
                *new_pairs.entry((*insert, rule_pair.1)).or_insert(0) += count;
            }
        }
    }
    return new_pairs;
}

fn solve_case<const STEPS: usize>((polymer, rules): &Input) -> usize {
    let chars: Vec<char> = polymer.chars().collect();
    let mut pairs = Pairs::new();
    for w in chars.windows(2) {
        *pairs.entry((w[0], w[1])).or_insert(0) += 1;
    }
    for _ in 0..STEPS {
        pairs = apply_rules(&pairs, rules);
    }
    let mut counts = [0usize; 26];
    for ((a, _), count) in &pairs {
        counts[*a as usize - 'A' as usize] += count;
    }
    counts[*chars.last().unwrap() as usize - 'A' as usize] += 1;
    let max = counts.iter().copied().max().unwrap();
    let min = counts.iter().copied().filter(|&c| c > 0).min().unwrap();
    return max - min;
}

fn main() {
    println!("Part 1");
    let example = parse("day14.example");
    aoc::expect_result!(1588, solve_case::<10>(&example));
    let input = parse("day14.input");
    aoc::expect_result!(2703, solve_case::<10>(&input));

    println!("Part 2");
    aoc::expect_result!(2188189693529, solve_case::<40>(&example));
    aoc::expect_result!(2984946368465, solve_case::<40>(&input));
}
